build/Templating.targets

```diff
@@ -0,0 +1,68 @@
+<Project>
+  <PropertyGroup>
+    <TemplatingProjectRoot>$(MSBuildThisFileDirectory)..\modules\Templating\</TemplatingProjectRoot>
+  </PropertyGroup>
+
+  <Target Name="BuildTemplates" DependsOnTargets="GeneratePropsFiles">
+    <PropertyGroup>
+      <GeneratedNoTimestampPackageVersionPropsPath>$(IntermediateDir)dependencies.notimestamp.props</GeneratedNoTimestampPackageVersionPropsPath>
+      <TemplateProjCommmonProperties>
+        RepositoryRoot=$(TemplatingProjectRoot);
+        DotNetRestoreSourcesPropsPath=$(GeneratedRestoreSourcesPropsPath);
+        BuildNumber=$(BuildNumber);
+        Configuration=$(Configuration);
+      </TemplateProjCommmonProperties>
+      <TemplateProjProperties>
+        $(TemplateProjCommmonProperties);
+        DotNetPackageVersionPropsPath=$(GeneratedPackageVersionPropsPath);
+        VersionMetadata=timestamped;
+      </TemplateProjProperties>
+      <TemplateProjNoTimestampProperties>
+        $(TemplateProjCommmonProperties);
+        DotNetPackageVersionPropsPath=$(GeneratedNoTimestampPackageVersionPropsPath);
+        VersionMetadata=notimestamp;
+      </TemplateProjNoTimestampProperties>
+    </PropertyGroup>
+
+    <!-- Produce regular, timestamped templates for pre-release builds -->
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
+             Targets="Restore;Compile;Package"
+             Properties="$(TemplateProjProperties)" />
+
+    <ItemGroup>
+      <TemplateArtifacts Include="$(TemplatingProjectRoot)artifacts\build\*" />
+    </ItemGroup>
+
+    <Copy SourceFiles="@(TemplateArtifacts)" DestinationFolder="$(BuildDir)" />
+
+    <!-- Gather artifact info as if this were a final build -->
+    <MSBuild Projects="$(MSBuildProjectFullPath)"
+             Targets="GetArtifactInfo"
+             Properties="RepositoryRoot=%(Repository.RootPath);Configuration=$(Configuration);BuildNumber=$(BuildNumber);IsFinalBuild=true"
+             ContinueOnError="WarnAndContinue">
+      <Output TaskParameter="TargetOutputs" ItemName="FinalBuildArtifactInfo" />
+    </MSBuild>
+
+    <ItemGroup>
+      <_NoTimestampPackages Include="@(ExternalDependency)" />
+      <_NoTimestampPackages Include="%(FinalBuildArtifactInfo.PackageId)" Version="%(FinalBuildArtifactInfo.Version)" Condition=" '%(FinalBuildArtifactInfo.ArtifactType)' == 'NuGetPackage' " />
+      <_NoTimestampPackages Include="Microsoft.AspNetCore.All" Version="$(PackageVersionNoTimestamp)" />
+    </ItemGroup>
+
+    <RepoTasks.GeneratePackageVersionPropsFile
```

Comment:
FYI @SteveSandersonMS  - this is the piece that updates version numbers. It spits out a file with MSBuild properties containing version overrides. It's passed into aspnet/Templating via the DotNetPackageVersionPropsPath property.